pkg/agent/packetcapture/packetcapture_controller_test.go

```diff
@@ -0,0 +1,597 @@
+		mockController:     controller,
+		crdClient:          crdClient,
+		crdInformerFactory: crdInformerFactory,
+		informerFactory:    informerFactory,
+	}
+}
+
+func addPodInterface(ifaceStore interfacestore.InterfaceStore, podNamespace, podName string, podIPs []string, podMac string, ofPort int32) {
+	containerName := k8s.NamespacedName(podNamespace, podName)
+	var ifIPs []net.IP
+	for _, ip := range podIPs {
+		ifIPs = append(ifIPs, net.ParseIP(ip))
+	}
+	mac, _ := net.ParseMAC(podMac)
+	ifaceStore.AddInterface(&interfacestore.InterfaceConfig{
+		IPs:                      ifIPs,
+		MAC:                      mac,
+		InterfaceName:            util.GenerateContainerInterfaceName(podName, podNamespace, containerName),
+		ContainerInterfaceConfig: &interfacestore.ContainerInterfaceConfig{PodName: podName, PodNamespace: podNamespace, ContainerID: containerName},
+		OVSPortConfig:            &interfacestore.OVSPortConfig{OFPort: ofPort},
+	})
+}
+
+func TestMultiplePacketCaptures(t *testing.T) {
+	defaultFS = afero.NewMemMapFs()
+	packetsDir := "/tmp/antrea/packetcapture/packets"
+	defaultFS.MkdirAll(packetsDir, 0755)
+	nameFunc := func(id int) string {
+		return fmt.Sprintf("pc-%d", id)
+	}
+	var objs []runtime.Object
+	for i := 0; i < 20; i++ {
+		objs = append(objs, genTestCR(nameFunc(i), int32(testCaptureNum)))
+	}
+	pcc := newFakePacketCaptureController(t, nil, objs)
+	pcc.sftpUploader = &testUploader{url: testFTPUrl}
+	stopCh := make(chan struct{})
+	defer close(stopCh)
+	pcc.crdInformerFactory.Start(stopCh)
+	pcc.crdInformerFactory.WaitForCacheSync(stopCh)
+	pcc.informerFactory.Start(stopCh)
+	pcc.informerFactory.WaitForCacheSync(stopCh)
+	go pcc.Run(stopCh)
+	assert.Eventually(t, func() bool {
+		items, err := pcc.crdClient.CrdV1alpha1().PacketCaptures().List(context.Background(), metav1.ListOptions{})
+		if err != nil {
+			return false
+		}
+		for _, result := range items.Items {
+			for _, cond := range result.Status.Conditions {
+				if cond.Type == crdv1alpha1.PacketCaptureComplete || cond.Type == crdv1alpha1.PacketCaptureFileUploaded {
+					if cond.Status == metav1.ConditionFalse {
+						return false
+					}
+				}
+			}
+		}
+		pcc.mutex.Lock()
+		if pcc.numRunningCaptures != 0 {
+			return false
+		}
+		pcc.mutex.Unlock()
+		return true
+	}, 5*time.Second, 50*time.Millisecond)
+	for i := 0; i < 20; i++ {
+		err := pcc.crdClient.CrdV1alpha1().PacketCaptures().Delete(context.TODO(), nameFunc(i), metav1.DeleteOptions{})
+		require.NoError(t, err)
+	}
+	assert.Eventually(t, func() bool {
+		pcc.mutex.Lock()
+		if len(pcc.captures) != 0 {
+			return false
+		}
+		pcc.mutex.Unlock()
+		return true
+	}, 2*time.Second, 20*time.Millisecond)
+
+}
+
+// TestPacketCaptureControllerRun was used to validate the whole run process is working. It doesn't wait for
+// the testing pc to finish. on sandbox env, no good solution to open raw socket.
+func TestPacketCaptureControllerRun(t *testing.T) {
+	pcs := []struct {
```

Comment:
We should add a test for the timeout case later on, by injecting a fake clock into the controller.

Reply:
it used have, but i found it's not very stable, i figure out it's the clock thing, but didn't know why it works sometimes. Also we have a timeout case in e2e.

Will mark this an an TODO item.